### PR TITLE
Modifying dotnet test to handle the new design for debug and making it unit testable

### PR DESCRIPTION
**Pull request by livarcocc**
Adding unit tests for the state machine of dotnet test. Implemented a state machine in dotnet test. 

Added a message handler for the version check message. Also introduced an IDotnetTest that handles state and handlers (the state machine).

Added the test discover start message handler and introduced a test runner.

Added the handler for the GetTestRunnerProcessInfo message. Also, modified dotnet test to have separate setter for the special message handlers for terminate and unknown messages and added a separate method to add new reporting channels to DotnetTest, so that it can handle the new listener for the test runner.

Added the test runner test discovery handlers.

Added handlers to deal with the test execution itself.

Updated dotnet-test program to use the message handlers during design time.

Added a test for the whole discover tests message flow.

Added a test for the run tests full message exchange.

### CODE REVIEW

**Comment by livarcocc** [comment]
cc @piotrpMSFT

File: src/dotnet/commands/dotnet-test/DotnetTest.cs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        private readonly IList<IReportingChannel> _channels;
+        private readonly IList<IDotnetTestMessageHandler> _messageHandlers;
+        private readonly ITestMessagesCollection _messages;
+
+        public IDotnetTestMessageHandler TestSessionTerminateMessageHandler { private get; set; }
+        public IDotnetTestMessageHandler UnknownMessageHandler { private get; set; }
+
+        public DotnetTestState State { get; private set; }
+
+        public string AssemblyUnderTest { get; }
```
Comment (piotrpMSFT) [inline src/dotnet/commands/dotnet-test/DotnetTest.cs]: not sure about the name here. It implied that this would be an Assembly, not a String.

Comment (livarcocc) [inline src/dotnet/commands/dotnet-test/DotnetTest.cs]: PathToAssemblyUnderTest?

File: src/dotnet/commands/dotnet-test/DotnetTest.cs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        {
+            _messageHandlers.Add(messageHandler);
+
+            return this;
+        }
+
+        public void StartHandlingMessages()
+        {
+            Message message;
+            while (_messages.TryTake(out message))
```
Comment (piotrpMSFT) [inline src/dotnet/commands/dotnet-test/DotnetTest.cs]: out param?

Comment (livarcocc) [inline src/dotnet/commands/dotnet-test/DotnetTest.cs]: It is the Try pattern. Either that or forcing null checks.

File: src/dotnet/commands/dotnet-test/MessageHandlers/TestDiscoveryStartMessageHandler.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            }
+
+            return nextState;
+        }
+
+        private void HandleMessage(IDotnetTest dotnetTest)
+        {
+            TestHostTracing.Source.TraceInformation("Starting Discovery");
+
+            DiscoverTests(dotnetTest);;
```
Comment (piotrpMSFT) [inline src/dotnet/commands/dotnet-test/MessageHandlers/TestDiscoveryStartMessageHandler.cs]: -= ;

Comment (livarcocc) [inline src/dotnet/commands/dotnet-test/MessageHandlers/TestDiscoveryStartMessageHandler.cs]: hu?!

**Comment by livarcocc** [comment]
cc @brthor @Sridhar-MS

File: test/dotnet-test.UnitTests/GivenThatWeWantToRunTests.cs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        [Fact]
+        public void Dotnet_test_handles_and_sends_all_the_right_messages()
+        {
+            var dotnetTestMessageScenario = new DotnetTestMessageScenario();
+
+            dotnetTestMessageScenario.TestRunnerMock
+                .Setup(t => t.GetProcessStartInfo())
+                .Returns(new ProcessStartInfo())
+//                .Callback(() => dotnetTestMessageScenario.TestRunnerChannelMock.Raise(
```
Comment (brthor) [inline test/dotnet-test.UnitTests/GivenThatWeWantToRunTests.cs]: commented out?

File: test/dotnet-test.UnitTests/GivenThatWeWantToRunTests.cs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+using Moq;
+using Newtonsoft.Json.Linq;
+using Xunit;
+
+namespace Microsoft.Dotnet.Tools.Test.Tests
+{
+    public class GivenThatWeWantToRunTests
+    {
+        [Fact]
+        public void Dotnet_test_handles_and_sends_all_the_right_messages()
```
Comment (brthor) [inline test/dotnet-test.UnitTests/GivenThatWeWantToRunTests.cs]: do we expect the "right" messages to change?

Would it be worth specifying what those messages should be?

File: test/dotnet-test.UnitTests/GivenThatWeWantToRunTests.cs (hunk 4 above)
Comment (livarcocc) [inline test/dotnet-test.UnitTests/GivenThatWeWantToRunTests.cs]: I forgot to delete it. Not needed.

File: test/dotnet-test.UnitTests/GivenThatWeWantToRunTests.cs (hunk 5 above)
Comment (livarcocc) [inline test/dotnet-test.UnitTests/GivenThatWeWantToRunTests.cs]: Well, there are quite a few of these messages. I believe the test should be the ultimate document of that. Also, the sequence is described here: https://github.com/dotnet/cli/issues/1376

File: src/dotnet/commands/dotnet-test/MessageHandlers/TestDiscoveryStartMessageHandler.cs (hunk 3 above)
Comment (piotrpMSFT) [inline src/dotnet/commands/dotnet-test/MessageHandlers/TestDiscoveryStartMessageHandler.cs]: remove the extra semicolon!

**Comment by piotrpMSFT** [comment]
:shipit:

File: src/dotnet/commands/dotnet-test/MessageHandlers/TestDiscoveryStartMessageHandler.cs (hunk 3 above)
Comment (brthor) [inline src/dotnet/commands/dotnet-test/MessageHandlers/TestDiscoveryStartMessageHandler.cs]: --;

**Comment by livarcocc** [comment]
@mmitche This has been waiting for windows for hours now.

**Comment by mmitche** [comment]
@livarcocc Yes, this is known.  There are significant slowdowns with the RAID array stemming from an unexplained VM restart last night.

**Pull request merged**